pkg/manifest/manifest.go

```diff
@@ -0,0 +1,132 @@
+package manifest
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/kubernetes/scheme"
+)
+
+// Manifest stores Kubernetes object in Raw from a file.
+// It stores the GroupVersionKind for the manifest.
+// Raw and Obj should always be kept in sync such that
+// each provides the same data but in different formats.
+// To ensure Raw and Obj are always in sync, they should not
+// be set directly but rather only be set by using NewManifest
+// constructor or by calling either method ManifestsFromFiles
+// or ParseManifests.
+type Manifest struct {
+	// OriginalFilename is set to the filename this manifest was loaded from.
+	// It is not guaranteed to be set or be unique, but we will set it when
+	// loading from disk to provide better debuggability.
+	OriginalFilename string
+
+	Raw []byte
+	GVK schema.GroupVersionKind
+
+	Obj *unstructured.Unstructured
+}
+
+// Create a manifest from bytes of a single kubernetes object to Manifest.
+// Setting originalFilename is optional since it is not guaranteed to be set
+// in the Manifest structure.
+func NewManifest(originalFilename string, in []byte) (*Manifest, error) {
+	var manifest Manifest
+	if err := manifest.UnmarshalJSON(in); err != nil {
+		return nil, err
+	}
+	manifest.OriginalFilename = originalFilename
+	return &manifest, nil
+}
+
+// UnmarshalJSON unmarshals bytes of single kubernetes object to Manifest.
+func (m *Manifest) UnmarshalJSON(in []byte) error {
+	if m == nil {
+		return errors.New("Manifest: UnmarshalJSON on nil pointer")
+	}
+
+	// This happens when marshalling
+	// <yaml>
+	// ---	(this between two `---`)
+	// ---
+	// <yaml>
+	if bytes.Equal(in, []byte("null")) {
+		m.Raw = nil
+		return nil
+	}
+
+	m.Raw = append(m.Raw[0:0], in...)
+	udi, _, err := scheme.Codecs.UniversalDecoder().Decode(in, nil, &unstructured.Unstructured{})
+	if err != nil {
+		return errors.Wrapf(err, "unable to decode manifest")
+	}
+	ud, ok := udi.(*unstructured.Unstructured)
+	if !ok {
+		return fmt.Errorf("expected manifest to decode into *unstructured.Unstructured, got %T", ud)
+	}
+
+	m.GVK = ud.GroupVersionKind()
+	m.Obj = ud.DeepCopy()
```

Comment:
why deepcopy?

Reply:
Must be required. I removed the DeepCopy and existing code gets a SIGSEGV nil pointer when trying to access Obj:

`src := fmt.Sprintf("the config map %s/%s", ms[i].Obj.GetNamespace(), ms[i].Obj.GetName())`

Reply:
That's odd. Of course I meant `m.Obj = ud`. Doesn't work?

Reply:
Yes, when I commented out line 65 'm.Obj = ud.DeepCopy()' it failed with SIGSEGV.

Reply:
Changed and does work